src/oc_erchef/apps/chef_index/c_src/cs_escape.c

```diff
@@ -0,0 +1,224 @@
+/* Copyright 2015 Chef Software, Inc.
+ *
+ * This file is provided to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License.  You may obtain
+ * a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ * cs_escape: Does text-replacement on select ASCII punctionation
+ * to avoid problems with CloudSearch's word-splitting rules.
+ *
+ */
+#include <string.h>
+#include <stdbool.h>
+#include "erl_nif.h"
+
+// The first 32 ASCII/UTF8 Characters are control characters.  We
+// don't expect to see them so we don't define replacements for them,
+// shortening the array.
+#define replacement_for(n) replacements[n - 33]
+
+static const char *replacements[] = {
+  "__EX__", // !
+  "__DQ__", // "
+  "__HS__", // #
+  "__DL__", // $
+  "__PC__", // %
+  "__AM__", // &
+  "'", // ' No replacement
+  "__OP__", // (
+  "__CP__", // )
+  "__ST__", // *
+  "__PL__", // +
+  "__CM__", // ,
+  "__DS__", // -
+  ".", // . No replacement
+  "__FS__", // Foward Slash
+  "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", // No replacement
+  ":", // :
+  "__SC__", // ;
+  "__LT__", // <
+  "__EQ__", // =
+  "__GT__", // >
+  "__QS__", // ?
+  "__AT__", // @
+  "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K",
+  "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V",
+  "W", "X", "Y", "Z", // No replacements
+  "__OB__", // [
+  "__BS__", // Back Slash
+  "__CB__", // ]
+  "__CA__", // ^
+  "_", // _
```

Comment:
My only concern here would be that actual data might have something like `__EQ__` in it, and then we hit that unintentionally because we search for `=`.